R/read-write.R

```diff
@@ -0,0 +1,428 @@
+  }
+  tryCatch(
+    dm_write_csv_impl(dm, csv_directory, zip = FALSE),
+    error = function(e) {
+      # remove directory in case it was created but there was an error
+      try(unlink(csv_directory, recursive = TRUE))
+      # keep error message and class though
+      abort(conditionMessage(e), .subclass = class(e))
+    }
+  )
+}
+
+#' @details `dm_read_csv()`: read a `dm` from a directory created using `dm_write_csv()`.
+#' @rdname dm-read-write
+#' @export
+dm_read_csv <- function(csv_directory) {
+  file_list <- list.files(csv_directory)
+  special_files <- c(
+    "___info_file_dm.csv",
+    "___coltypes_file_dm.csv",
+    "___pk_file_dm.csv",
+    "___fk_file_dm.csv"
+  )
+  if (!all(special_files %in% file_list)) {
+    abort_files_or_sheets_missing(setdiff(special_files, intersect(file_list, special_files)), csv_directory)
+  }
+
+  def_base <- readr::read_csv(
+    file.path(csv_directory, "___info_file_dm.csv"),
+    col_types = "ccc"
+  )
+  table_names <- def_base$table
+  col_class_info <- readr::read_csv(
+    file.path(csv_directory, "___coltypes_file_dm.csv"),
+    col_types = "ccc"
+  ) %>%
+    mutate(readr_class = readr_translate(class)) %>%
+    group_by(table) %>%
+    summarize(readr_class = paste0(readr_class, collapse = ""))
+  # sort column classes according to column def_base$table
+  col_class_sorted <- col_class_info$readr_class[order(match(col_class_info$table, table_names))]
+
+  pk_info <- readr::read_csv(
+    file.path(csv_directory, "___pk_file_dm.csv"),
+    col_types = "cc"
+  )
+
+  fk_info <- readr::read_csv(
+    file.path(csv_directory, "___fk_file_dm.csv"),
+    col_types = "ccci"
+  )
+
+  table_files <- setdiff(
+    list.files(csv_directory),
+    c("___info_file_dm.csv", "___pk_file_dm.csv", "___fk_file_dm.csv", "___coltypes_file_dm.csv")
+  )
+  # sort table_files according to column def_base$table
+  table_files_sorted <- table_files[order(match(table_files, paste0(table_names, ".csv")))]
+
+  table_tibble <- map2(
+    file.path(csv_directory, table_files_sorted),
+    col_class_sorted,
+    ~ readr::read_csv(file = .x, col_types = .y)
+  ) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data") %>%
+    # empty subset is necessary because
+    # https://www.tidyverse.org/blog/2018/12/readr-1-3-1/#tibble-subclass
+    mutate(data = map(data, `[`))
+
+  make_dm(def_base, table_names, table_tibble, pk_info, fk_info)
+}
+
+#' @inheritParams dm_write_csv
+#' @param zip_file_path
+#' For `dm_write_zip`: The file path to the `zip`-file to write; defaults to `dm.zip`.
+#'
+#' For `dm_read_zip`: The file path to the `zip`-file to read the `dm` from.
+#' @param overwrite Logical, default: `FALSE`. In case the file already exists, should it be overwritten?
+#' @details `dm_write_zip()`: write a `dm` to a `zip`-file containing a collection of `csv`-files.
+#' @rdname dm-read-write
+#' @export
+dm_write_zip <- function(dm, zip_file_path = "dm.zip", overwrite = FALSE) {
```

Comment:
Not sure I like the default file name.

Reply:
Why? or to ask differently, do you have another name in mind? I guess the same is valid for `dm_write_xlsx(dm, xlsx_file_path = "dm.xlsx")`?

Reply:
Do we even need a default?

Reply:
we don't really, removing.